### ISSUE TITLE
A new approach to writing the Dockerfile

### ISSUE THREAD

**Issue by karthik** [karthik/holepunch#8]
@annakrystali writes:

> One thing that’s “maybe missing” is a runtime.txt file, but then I’ve also heard that using a Dockerfile means that runtime.txt is ignored? Indeed this is my main conceptual hole about the Dockerfile + DESCRIPTION approach currently, ie how can we capture specific versions of dependencies? I've tried the package (== version) approach in the DESCRIPTION but that didn't seem to work. This was actually on my top priority to try and figure out so if you have more info on this it would be great to hear!

This is indeed correct.

So I have a better proposal for writing the Dockerfile and would like some feedback.

There is a new function called `last_modification_date` that will capture the last date any file on the project was touched and use that as the date to grab the R version and associated packages. This is useful if you are turning binder on an old project, where you don't accidentally want to use the current R version + latest packages.

So with `last_modification_date` + R version for that date, we write a Dockerfile that does:

```
FROM rocker/binder:latest
LABEL maintainer="karthik"
USER root
COPY . ${HOME}
RUN chown -R ${NB_USER} ${HOME}
## Become normal user again
USER ${NB_USER}
RUN wget https://github.com/karthik/delete-asap/raw/master/DESCRIPTION && R -e "options(repos = list(CRAN = 'http://mran.revolutionanalytics.com/snapshot/{DATE}/'); 'devtools::install_deps()"
```

Thoughts on this @annakrystali and @benmarwick?

**Comment by karthik** [comment]
cc @annakrystalli (tagging you correctly this time)

**Comment by benmarwick** [comment]
Wow, yes, that is a very thoughtful addition. I love it! So when I run `write_dockerfile()` I will get a Dockerfile makes an image with the R pkgs from the date that I last updated any file in my project.

I wonder if we can also call  the base image at a specific version of R, is this possible? e.g.  `FROM rocker/binder:3.4.0` Not sure if that's better than a specific date because pkgs can change a lot in 6 months. The only upshot might be that we get compiled pkgs with the base image, but if we get pkgs from MRAN we need to wait to compile, is that right?

**Comment by karthik** [comment]
> I wonder if we can also call the base image at a specific version of R, is this possible? e.g. FROM rocker/binder:3.4.0

Yes indeed! Thanks for catching that mistake. I forgot to add that in my mock example above.  The correct version would be:

```
FROM rocker/tidyverse:{R_VERSION}
```
So this will account both for the version of R and also the packages/binaries for that version. I now need some sort of look up table to match date ranges to versions of R and then I'll be able to push this update.

**Comment by annakrystalli** [comment]
😃Really like this solution! 🙌

My one suggestion for more robustness would be to be able to manually override the date. I can imagine situations where perhaps you've got an old project but have recently done minor bug fixes or simple updates to a README file that you might not want to dictate the "computational date" of the project. In this case being able to manually supply the {DATE} field could be quite useful.

On a more general note though, there is no way to capture the actual versions of packages used is there? In my mind the only weakness in the MRAN snapshot approach is that it assumes that the user has up to date versions of all packages involved installed which, might be likely but not necessarily guaranteed.

**Comment by karthik** [comment]
> My one suggestion for more robustness would be to be able to manually override the date.

Awesome! I am working on that exact thing right now! A base or date in the function argument will override this automatic date choice.

> On a more general note though, there is no way to capture the actual versions of packages used is there?

The MRAN date maps to the package versions. So for whatever date you set,  packages on CRAN at that date will be the version chosen. Unfortunately what you want isn't possible in the R ecosystem.

**Comment by karthik** [comment]
Added some comments here that I'll code up shortly: 

https://github.com/karthik/holepunch/blob/master/R/write_dockerfile.R#L24-L31

**Comment by karthik** [comment]
This should be resolved in the current version (`0.0.9.9000`).

You can try:

```r
write_dockerfile(maintainer = "karthik")
# → Setting R version to 3.6.0
# ✔ Dockerfile written to .binder/Dockerfile
```
This works by picking the last modified date of files in the project.

You can override this by setting a date manually

```r
write_dockerfile(maintainer = "karthik", r_date = "2018-06-25")
# → Setting R version to 3.5.0
# ✔ Dockerfile written to .binder/Dockerfile
```

Let's say you've got a base with everything installed, you can override with your own base:

```r
write_dockerfile(maintainer = "karthik", base = "karthik/bbase")
# → Setting R version to 3.6.0
# ✔ Dockerfile written to .binder/Dockerfile
```
The `Dockerfile will now look like this:

```r
FROM karthik/bbase
LABEL maintainer='karthik'
...
```

You can set a base and a date if necessary:

```r
write_dockerfile(maintainer = "karthik", base = "karthik/bbase", r_date = "2015-12-15")
# → Setting R version to 3.2.3
# ✔ Dockerfile written to .binder/Dockerfile
```

But I imagine bad things will happen here if the R version doesn't match the one in your base.


Does this solve all use cases?

**Comment by annakrystalli** [comment]
#### Amazing work!! This is so cool! 🎉
![](https://media1.tenor.com/images/a33bfcc001cf968ada2accb367a14896/tenor.gif?itemid=5479113)

A few additional comments:

- The user now defaults to [`"karthik"`](https://github.com/annakrystalli/testdrive-holepunch/blob/7cff78396bb17ca58b66be3cd0353339617dae37/.binder/Dockerfile) whereas before it was [correctly guessing `"Anna"`](https://github.com/annakrystalli/testdrive-holepunch/blob/12201845801dc0c74b041303b91199eb39d35216/.binder/Dockerfile) for me. Now have to remember to set this manually.

- I tried to use it with the tutorial material and set the date to `"2017-01-01"`. It seems to set the version of R correctly, (3.3.2). However when I try to `build_binder()` it fails with the message: `"manifest for rocker/binder:3.3.2 not found"`.  There's clearly a tag for 3.3.2 on [rocker/binder](https://github.com/rocker-org/binder/tags) on GitHub but not on [dockerhub](https://hub.docker.com/r/rocker/binder/tags) 🤷‍♀️. One for @cboettig to be aware of perhaps?

- Using `"2019-01-01"` and `"2018-06-01"` works like a treat 🤩

- In the `write_dockerfile()` docs, it might be a good idea to specify the format required for the date argument (ie ISO).
- Also might be a good idea to note in `write_dockerfile()` (and perhaps even in `build_binder()` docs) that changes in the Dockerfile will only take effect once it's pushed to GitHub (ie make clear that it's the file on GitHub that is used to build the image). I also found myself needing to restart my R session in order for the Dockerfile to appear as modified in my git tab and therefore allow me to commit & push it. Admittedly that's an Rstudio thing but perhaps could be added to the note about ensuring to commit and push Dockerfile for it to take effect.

- Finally very minor comment re: current print out of the build log messages. At the minute they are just being printed off as a list which can be hard to follow (see below). A completely non-essential enhancement could be to translate those into visually more accessible format. But certainly not a priority. 

```r
$message
[1] "\033[91m** building package indices\n\033[0m"

$phase
[1] "building"

$message
[1] "\033[91m** installing vignettes\n\033[0m"

$phase
[1] "building"

$message
[1] "\033[91m** testing if installed package can be loaded\n\033[0m"

$phase
[1] "building"

$message
[1] "\033[91m* DONE (dataRetrieval)\n\033[0m"

$phase
[1] "building"

$message
[1] "> \n> \n"

$phase
[1] "building"

$message
[1] "Removing intermediate container 987de49aea05\n"

$phase
[1] "building"

$message
[1] " ---> 0802aac1eeb6\n"

$phase
[1] "building"

$message
[1] "{\"aux\": {\"ID\": \"sha256:0802aac1eeb6e09308514c58f2c1dab333e58617b96892a0747a9631def4e4aa\"}}"

$phase
[1] "building"

$message
[1] "[Warning] One or more build-args [NB_USER NB_UID] were not consumed\n"

$phase
[1] "building"

$message
[1] "Successfully built 0802aac1eeb6\n"

```

**Comment by karthik** [comment]
Thanks @annakrystalli! I opened these as separate issues to address.

**Comment by karthik** [comment]
Closing as most of these are resolved (or have related open issues)
